cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/worker/sidecar/ArtifactLocalizer.java

```diff
@@ -0,0 +1,236 @@
+
+    if (lastModifiedTimestamp != null) {
+      LOG.debug("Found existing local version with timestamp {}", lastModifiedTimestamp);
+      url += String.format("&lastModified=%d", lastModifiedTimestamp);
+    }
+
+    Long finalLastModifiedTimestamp = lastModifiedTimestamp;
+    String finalUrl = url;
+    Location newJarLocation = Retries
+      .callWithRetries(() -> fetchArtifact(artifactId, finalUrl, finalLastModifiedTimestamp), retryStrategy);
+
+    // If we didn't have a version that was previous cached then we're done and we can return the location
+    if (lastModifiedTimestamp == null) {
+      return newJarLocation;
+    }
+
+    // This means we already have a jar but its out of date, we should delete the jar and the unpacked directory
+    Location oldJarLocation = getArtifactJarLocation(artifactId, lastModifiedTimestamp);
+    LOG.debug("Got new location as {} with old location as {}", newJarLocation, oldJarLocation);
+    if (!newJarLocation.equals(oldJarLocation)) {
+      Location oldUnpackLocation = getUnpackLocalPath(artifactId, lastModifiedTimestamp);
+      LOG.debug("Deleting previously cached jar");
+      try {
+        oldJarLocation.delete();
+        oldUnpackLocation.delete(true);
+      } catch (IOException e) {
+        //Catch and log the exception, this should not cause the operation to fail
+        LOG.warn("Failed to delete old cached jar for artifact {} version {}: {}", artifactId.getArtifact(),
+                 artifactId.getVersion(), e);
+      }
+    }
+    return newJarLocation;
+  }
+
+  /**
+   * Gets the location on the local filesystem for the directory that contains the unpacked artifact. This method
+   * handles fetching, caching and unpacking the artifact.
+   *
+   * @param artifactId The ArtifactId of the artifact to fetch and unpack
+   * @return The Local Location of the directory that contains the unpacked artifact files
+   * @throws ArtifactNotFoundException if the given artifact does not exist
+   * @throws IOException if there was an exception while fetching, caching or unpacking the artifact
+   * @throws Exception if there was an unexpected error
+   */
+  public Location getAndUnpackArtifact(ArtifactId artifactId) throws Exception {
+    LOG.debug("Unpacking artifact {}", artifactId);
+    Location jarLocation = getArtifact(artifactId);
+    Location unpackDir = getUnpackLocalPath(artifactId, Long.valueOf(jarLocation.getName().split("\\.")[0]));
+    LOG.debug("Got unpack directory as {}", unpackDir);
+    if (!unpackDir.exists()) {
+      LOG.debug("Unpack directory doesnt exist, creating it now");
+      BundleJarUtil.unJar(jarLocation, new File(unpackDir.toURI()));
+    }
+    return unpackDir;
+  }
+
+  private Location fetchArtifact(ArtifactId artifactId, String url,
+                                 Long lastModifiedTimestamp) throws IOException, ArtifactNotFoundException {
+    HttpURLConnection urlConn = remoteClient.openConnection(HttpMethod.GET, url);
```

Comment:
Shouldn't the `lastModifiedTimestamp` be provided to the  get call?

Reply:
It was added to the URL before this method is called. But after switching it to use the header it is added here.